pyquil/api/_quantum_computer.py

```diff
@@ -189,6 +431,81 @@ def run_symmetrized_readout(self, program: Program, trials: int) -> np.ndarray:
+        measure in the Z basis. Then the Confusion matrix for the readout error is specified by
+        the probabilities
+
+            p_{j|i} := Pr(measured = j | prepared = i )
+
+        If p_{0|0} = p_{1|1} = 1 then there is no readout error.
+        If p_{0|0} = p_{1|1} = 1 - epsilon then the readout error is symmetric.
+        If p_{0|0} != p_{1|1}  then the readout error is asymmetric.
+
+        If your quantum computer has this kind of asymmetric readout error then
+        `qc.run_symmetrized_readout` will symmetrize the readout error.
+
+        The readout error above is only asymmetric on a single bit. In practice the confusion
+        matrix on n bits need not be symmetric, e.g. for two qubits p_{ij|ij} != 1 - epsilon for
+        all i,j. In these situations a more sophisticated means of symmetrization is needed; and
+        we use orthogonal arrays (OA).
+
+        The symmetrization types are specified by an int; the types available are:
+        -1 -- exhaustive symmetrization uses every possible combination of flips
+         0 -- trivial that is no symmetrization
+         1 -- symmetrization using an OA with strength 1
+         2 -- symmetrization using an OA with strength 2
+         3 -- symmetrization using an OA with strength 3
+        By default a strength 3 OA is used; this ensures expectations of the form <b_k b_j b_i>
+        for bits any bits i,j,k will have symmetric readout errors. As a strength 3 OA is also a
+        strength 2 and strength 1 OA it also ensures <b_j b_i> and <b_i> have symmetric readout
+        errors for any bits b_j and b_i.
+
+        :param program: The program to run symmetrized readout on.
+        :param trials: The minimum number of times to run the program.
+        :param symm_type: the type of symmetrization
+        :param meas_qubits: An advanced feature. The groups of measurement qubits. Only these
+            qubits will be symmetrized over, even if the program acts on other qubits.
+        :return: A numpy array of shape (trials, len(ro-register)) that contains 0s and 1s.
+        """
+        if type(symm_type) is not int:
+            raise ValueError("Symmetrization options are indicated by an int. See "
+                             "the docstrings for more information.")
+
+        if trials % 2 != 0:
+            raise ValueError("Using symmetrized measurement functionality requires that you "
+                             "take an even number of trials.")
+
+        if meas_qubits is None:
+            meas_qubits = list(program.get_qubits())
+
+        if trials <= len(meas_qubits):
+            trials = _next_power_of_2(2 * len(meas_qubits) + 1)
+            warnings.warn('Number of trials was too low, it is now ' + str(trials))
+
+        sym_programs, flip_arrays = symmetrization(program, meas_qubits, symm_type)
+        # This will be 1 or a power of two
+        num_sym_progs = len(sym_programs)
+        num_shots_per_prog = trials // num_sym_progs
```

Comment:
I'm still confused by the number of trials here. You ensure that trials is even and greater than the number of meas_qubits, but consider 
```python
len(meas_qubits) = 6
trials = 10
symm_type = -1
num_sym_progs = 2 ** 6
num_shots_per_prog = trials // num_sym_progs = 10 / 2 ** 6 = 0
```

Is the check that trials is even supposed to be making trials the next power of two? Even so the example above would fail. 

Further wouldn't we want to try to optimize the number of shots here? So for ```len(measure_qubits) = 6``` and initial ```trials=8```:
1) with a one-design we  keep ```trials = 8``` and go with ```num_shots_per_prog = 5```
2) with a two-design we keep ```trials = 8``` and set ```num_shots_per_prog = 1```
3) for three-design we need ```trials = 16``` and ```num_shots_per_prog = 1``` 
4) for exhaustive we need ```trials = 2**6```

Reply:
Ok I just change the code to check if the minimum number of trials is achieve on a per symm_type basis. I also edited the docstrings to say that typically we want many more trials than the minimum.

Take a look now and let me know what you think.

Reply:
So the changes I am just about to push do the following. Given `progRAM = Program([I(x) for x in range(0,6)])`

```python
qc.run_symmetrized_readout_new(progRAM, trials=10, symm_type=-1)

UserWarning: Number of trials was too low, it is now 64
# Below here is printed out for debugging purposes
trials 64
sym progs 64
(floor division) num_shots_per_prog 1
```
and the output for the cases you give above is
(1) 
```python
trials 8
sym progs 2
(floor division) num_shots_per_prog 4
```
(2) 
```python
trials 8
sym progs 8
(floor division) num_shots_per_prog 1
```

(3) 
```
UserWarning: Number of trials was too low, it is now 16
trials 16
sym progs 16
(floor division) num_shots_per_prog 1
```

(4) 
```python
UserWarning: Number of trials was too low, it is now 64
trials 64
sym progs 64
(floor division) num_shots_per_prog 1
```

Reply:
This sounds good to me. As you suggested in person it is probably good to warn the user if 
```num_shots_per_program * len(symm_progrs) < trials```
We should probably document well which number of trials for various options won't be changed.